### PR TITLE
Fix mypy errors attributed to `pytorch_lightning.loggers.logger.py`

### PR DESCRIPTION
**Pull request by JustinGoheen**
## What does this PR do?

Fix mypy errors attributed to pytorch_lightning.loggers.logger.py for issue #13445

Fixes #\<issue_number>

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: src/pytorch_lightning/loggers/logger.py
Comment (awaelchli) [inline src/pytorch_lightning/loggers/logger.py:34]:
```suggestion
if TYPE_CHECKING:
```

File: src/pytorch_lightning/loggers/logger.py
Comment (awaelchli) [inline src/pytorch_lightning/loggers/logger.py:42]: We shouldn't annotate self. self has always the pseudo type "Self"

**Review by awaelchli — APPROVED** [review]

File: src/pytorch_lightning/loggers/logger.py
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/logger.py:42]: When I remove that annotation, mypy complains:

`pytorch_lightning/loggers/logger.py:42: error: Function is missing a type annotation for one or more arguments`.

Rather than annotate Self, should I comment with `# type: ignore[no-untyped-def]` to exclude the error?

**Review by akihironitta — APPROVED** [review]
@JustinGoheen Other than my comments below, it looks good to me!

> Also, the following PRs are waiting on checks, is there anything more I need to do before the remaining checks can run?

Let's wait for reviews from others!

**Review by otaj — APPROVED** [review]

**Comment by otaj** [comment]
Oh, things started failing as `ExperimentWriter` is undefined. I'll take a closer look at it on Monday

File: src/pytorch_lightning/loggers/logger.py
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/logger.py:165]: The current code on default branch is

```python
    def log_graph(self, model: "pl.LightningModule", input_array=None) -> None:
        """Record model graph.

        Args:
            model: lightning model
            input_array: input passes to `model.forward`
        """
        pass
```

Originally, mypy complained that there were args without annotations; so I annotated with `input_array: Tensor = None`, and mypy raised an error that Tensor cannot be None; I then annotated as `input_array: Optional[Tensor] = None`, and the error was resolved.

File: src/pytorch_lightning/loggers/logger.py
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/logger.py:39]: I had annotated, and then it was decided to not annotate self. Please see https://github.com/Lightning-AI/lightning/pull/13541#discussion_r914449104. Maybe this is something to table internally so a convention can be set?

File: src/pytorch_lightning/loggers/logger.py
Comment (JustinGoheen) [reply src/pytorch_lightning/loggers/logger.py:163]: I'm not sure why mypy doesn't flag this. Committing the suggestion.

**Comment by JustinGoheen** [comment]
@otaj I had set a conditional import of ExperimentWriter; I removed that condition in favor of a hardcoded import.

**Comment by otaj** [comment]
@JustinGoheen Great :)

**Pull request merged**
